embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/EmbraceDeliveryCacheManager.kt

```diff
@@ -50,6 +51,7 @@ internal class EmbraceDeliveryCacheManager(
     // The session id is used as key for this map
     // This list is initialized when getAllCachedSessions() is called.
     private val cachedSessions = mutableMapOf<String, CachedSession>()
+    private val allowPeriodicSessionCaching = AtomicBoolean(true)
```

Comment:
It is probably worth altering `SessionOrchestratorImpl` so that `processEndMessage` is invoked after the periodic session caching has stopped. The current implementation has a race condition where the periodic caching could happen after the session crash snapshot was written to disk. In this scenario the collections could already be cleaned which would explain why some payloads contained empty span arrays.

I think that moving the call order in `SessionOrchestratorImpl` so periodic caching is always suspended as the first operation would be enough to avoid the need for this check.

https://github.com/embrace-io/embrace-android-sdk/blob/6.3.2/embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/session/orchestrator/SessionOrchestratorImpl.kt#L200

Reply:
THAT what was I was looking for! We try to cache a session that is complete, and we should never do that. Nice. I'll have a look